// @jimp-native/plugin-resize-napi/es/index.js
import { throwError } from "@jimp/utils";
import { ensureInteger, getAddonReleaseVersion, cppCallbackWrapper, wrapAsync } from "@jimp-native/utils-ts";
var addon = require("../build/".concat(getAddonReleaseVersion(), "/plugin-resize-napi.node"));
var RESIZE_METHOD_MAP = {
  nearestNeighbor: 0,
  nearestNeighbour: 0,
  bilinearInterpolation: 1,
  bicubicInterpolation: 2,
  hermiteInterpolation: 3,
  bezierInterpolation: 4,
  defaultInterpolation: 99
};
function resize(width, height) {
  var mode = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "defaultInterpolation";
  var cb = arguments.length > 3 ? arguments[3] : undefined;
  if (typeof mode === "function" && typeof cb === "undefined") {
    cb = mode;
    mode = "defaultInterpolation";
  }
  if (width === this.constructor.AUTO) {
    width = this.bitmap.width * (height / this.bitmap.height);
  }
  if (height === this.constructor.AUTO) {
    height = this.bitmap.height * (width / this.bitmap.width);
  }
  if (width <= 0) {
    return throwError.call(this, "Width must be greater than zero", cb);
  }
  if (height <= 0) {
    return throwError.call(this, "Height must be greater than zero", cb);
  }
  var modeNumber = RESIZE_METHOD_MAP[mode];
  if (typeof modeNumber !== "number") {
    modeNumber = 99;
  }
  width = ensureInteger(width);
  height = ensureInteger(height);
  var targetBitmapData = Buffer.alloc(width * height * 4);
  try {
    var originalBitmap = this.bitmap.data;
    var originalWidth = this.bitmap.width;
    var originalHeight = this.bitmap.height;
    this.bitmap.data = targetBitmapData;
    this.bitmap.width = width;
    this.bitmap.height = height;
    addon.resize(originalBitmap, originalWidth, originalHeight, targetBitmapData, width, height, modeNumber, cppCallbackWrapper(this, cb));
    return this;
  } catch (err) {
    return throwError.call(this, err, cb);
  }
}
var plugin = function plugin() {
  return {
    constants: {
      RESIZE_NEAREST_NEIGHBOR: "nearestNeighbor",
      RESIZE_BILINEAR: "bilinearInterpolation",
      RESIZE_BICUBIC: "bicubicInterpolation",
      RESIZE_HERMITE: "hermiteInterpolation",
      RESIZE_BEZIER: "bezierInterpolation"
    },
    "class": {
      resize: resize,
      resizeAsync: wrapAsync(resize)
    }
  };
};
export default plugin;
