// @jimp-native/plugin-color-napi/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _utils = require("@jimp/utils");
var _pluginColor = _interopRequireDefault(require("@jimp/plugin-color"));
var _utilsTs = require("@jimp-native/utils-ts");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
var addon = require("../build/".concat((0, _utilsTs.getAddonReleaseVersion)(), "/plugin-color-napi.node"));
var PIXELATE_KERNEL = [[1 / 16, 2 / 16, 1 / 16], [2 / 16, 4 / 16, 2 / 16], [1 / 16, 2 / 16, 1 / 16]];
var greyscale = function greyscale(cb) {
  try {
    addon.greyscale(this.bitmap.data, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var brightness = function brightness(_brightness, cb) {
  try {
    addon.brightness(this.bitmap.data, _brightness, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var opacity = function opacity(_opacity, cb) {
  try {
    addon.opacity(this.bitmap.data, _opacity, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var opaque = function opaque(cb) {
  try {
    addon.opaque(this.bitmap.data, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var fade = function fade(f, cb) {
  return opacity.bind(this)(1 - f, cb);
};
var contrast = function contrast(_contrast, cb) {
  try {
    addon.contrast(this.bitmap.data, _contrast, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var posterize = function posterize(multiplier, cb) {
  try {
    addon.posterize(this.bitmap.data, multiplier, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var sepia = function sepia(cb) {
  try {
    addon.sepia(this.bitmap.data, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var convolution = function convolution(kernel) {
  var edgeHandling = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _utilsTs.EdgeHandling.EDGE_EXTEND;
  var cb = arguments.length > 2 ? arguments[2] : undefined;
  if (typeof edgeHandling === "function") {
    cb = edgeHandling;
    edgeHandling = _utilsTs.EdgeHandling.EDGE_EXTEND;
  }
  try {
    addon.convolution(this.bitmap.data, this.getWidth(), this.getHeight(), kernel, edgeHandling, 0, 0, this.getWidth(), this.getHeight(), 1, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var convolute = function convolute(kernel, x, y, w, h, cb) {
  if (typeof x === "function") {
    cb = x;
    x = 0;
  }
  try {
    var _x;
    addon.convolution(this.bitmap.data, this.getWidth(), this.getHeight(), kernel, _utilsTs.EdgeHandling.EDGE_EXTEND, (0, _utilsTs.ensureInteger)((_x = x) !== null && _x !== void 0 ? _x : 0), (0, _utilsTs.ensureInteger)(y !== null && y !== void 0 ? y : 0), (0, _utilsTs.ensureInteger)(w !== null && w !== void 0 ? w : this.getWidth()), (0, _utilsTs.ensureInteger)(h !== null && h !== void 0 ? h : this.getHeight()), 1, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var pixelate = function pixelate(size, x, y, w, h, cb) {
  var _x2, _y;
  if (typeof x === "function") {
    cb = x;
    x = null;
    y = null;
    w = null;
    h = null;
  }
  x = (_x2 = x) !== null && _x2 !== void 0 ? _x2 : 0;
  y = (_y = y) !== null && _y !== void 0 ? _y : 0;
  w = w || this.bitmap.width - (0, _utilsTs.ensureInteger)(x);
  h = h || this.bitmap.height - (0, _utilsTs.ensureInteger)(y);
  try {
    addon.convolution(this.bitmap.data, this.getWidth(), this.getHeight(), PIXELATE_KERNEL, _utilsTs.EdgeHandling.EDGE_EXTEND, (0, _utilsTs.ensureInteger)(x), (0, _utilsTs.ensureInteger)(y), (0, _utilsTs.ensureInteger)(w), (0, _utilsTs.ensureInteger)(h), size, (0, _utilsTs.cppCallbackWrapper)(this, cb));
    return this;
  } catch (err) {
    return _utils.throwError.call(this, err, cb);
  }
};
var plugin = function plugin() {
  return _objectSpread(_objectSpread({}, (0, _pluginColor["default"])()), {}, {
    // Merge with original implementation as our C++ coverage of color is not complete
    greyscale: greyscale,
    greyscaleAsync: (0, _utilsTs.wrapAsync)(greyscale),
    grayscale: greyscale,
    grayscaleAsync: (0, _utilsTs.wrapAsync)(greyscale),
    brightness: brightness,
    brightnessAsync: (0, _utilsTs.wrapAsync)(brightness),
    opacity: opacity,
    opacityAsync: (0, _utilsTs.wrapAsync)(opacity),
    opaque: opaque,
    opaqueAsync: (0, _utilsTs.wrapAsync)(opaque),
    fade: fade,
    fadeAsync: (0, _utilsTs.wrapAsync)(fade),
    contrast: contrast,
    contrastAsync: (0, _utilsTs.wrapAsync)(contrast),
    posterize: posterize,
    posterizeAsync: (0, _utilsTs.wrapAsync)(posterize),
    sepia: sepia,
    sepiaAsync: (0, _utilsTs.wrapAsync)(sepia),
    convolution: convolution,
    convolutionAsync: (0, _utilsTs.wrapAsync)(convolution),
    convolute: convolute,
    convoluteAsync: (0, _utilsTs.wrapAsync)(convolute),
    pixelate: pixelate,
    pixelateAsync: (0, _utilsTs.wrapAsync)(pixelate)
  });
};
var _default = plugin;
exports["default"] = _default;
module.exports = exports.default;
